obelisk/cpp/obelisk_cpp/include/obelisk_node.h

```diff
@@ -0,0 +1,820 @@
+        class ObeliskPublisherBase {
+          public:
+            virtual void OnActivate()   = 0;
+            virtual void OnDeactivate() = 0;
+            virtual void Release()      = 0;
+
+          protected:
+          private:
+        };
+
+        template <typename MessageT> class ObeliskPublisher : public ObeliskPublisherBase {
+          public:
+            ObeliskPublisher(typename rclcpp_lifecycle::LifecyclePublisher<MessageT>::SharedPtr publisher)
+                : publisher_(publisher) {}
+
+            void OnActivate() override { publisher_->on_activate(); }
+
+            void OnDeactivate() override { publisher_->on_deactivate(); }
+
+            void Release() override { publisher_.reset(); }
+
+            typename rclcpp_lifecycle::LifecyclePublisher<MessageT>::SharedPtr GetPublisher() { return publisher_; }
+
+          protected:
+          private:
+            typename rclcpp_lifecycle::LifecyclePublisher<MessageT>::SharedPtr publisher_;
+        };
+
+        struct ObeliskPublisherInfo {
+            std::string ros_param;
+            std::string msg_type;
+            std::function<std::shared_ptr<ObeliskPublisherBase>(const std::string& config_str)> creator;
+        };
+
+        // ------------------------------------- //
+        // ------------ Subscribers ------------ //
+        // ------------------------------------- //
+        class ObeliskSubscriptionBase {
+          public:
+            virtual void Release() = 0;
+
+          protected:
+          private:
+        };
+
+        template <typename MessageT> class ObeliskSubscription : public ObeliskSubscriptionBase {
+          public:
+            ObeliskSubscription(typename rclcpp::Subscription<MessageT>::SharedPtr subscription)
+                : subscription_(subscription) {}
+
+            void Release() override { subscription_.reset(); }
+
+            typename rclcpp::Subscription<MessageT>::SharedPtr GetSubscription() { return subscription_; }
+
+          protected:
+          private:
+            typename rclcpp::Subscription<MessageT>::SharedPtr subscription_;
+        };
+
+        struct ObeliskSubscriptionInfo {
+            std::string ros_param;
+            std::string msg_type;
+            std::function<std::shared_ptr<ObeliskSubscriptionBase>(const std::string& config_str)> creator;
+        };
+
+        // ------------------------------------- //
+        // --------------- Timers -------------- //
+        // ------------------------------------- //
+        struct ObeliskTimerInfo {
+            std::string ros_param;
+            std::function<rclcpp::TimerBase::SharedPtr(const std::string&)> creator;
+        };
+
+    } // namespace internal
+
+    /**
+     * ObeliskNode class.
+     * A lifecycle node whose publishers and subscribers can only publish and
+     * subscribe to Obelisk messages.
+     */
+    class ObeliskNode : public rclcpp_lifecycle::LifecycleNode {
```

Comment:
once other details match up, we should try to also match the docstrings between the cpp and python version